records_mover/url/optimizer/__init__.py

```diff
@@ -0,0 +1,349 @@
+
+        if not optimized_directory.writable():
+            logger.info(f"{optimized_directory} is not writable")
+            return None
+
+        return optimized_directory
+
+    @contextmanager
+    def _optimize_temp_locations_for_gcp_data_transfer(self,
+                                                       temp_first_loc: 'S3DirectoryUrl',
+                                                       temp_second_loc: 'GCSDirectoryUrl') ->\
+            Iterator[Tuple[BaseDirectoryUrl, BaseDirectoryUrl]]:
+        #
+        # GCP data transfer is great, but has the limitations:
+        #
+        # 1) It only works from GCS -> GCS or S3 -> GCS - like the
+        # roach motel, you can check in but you can't check out.
+        #
+        # 2) You can't specify a different destination location
+        # directory in the GCS bucket than in your source bucket
+        #
+        # So let's make sure that if at all possible, we use the same
+        # directory.
+        optimized_temp_first_loc = self.try_swapping_bucket_path(temp_first_loc,
+                                                                      temp_second_loc)
+        if optimized_temp_first_loc is not None:
+            # TODO: Delete after
+            yield (optimized_temp_first_loc, temp_second_loc)
+            return
+        optimized_temp_second_loc = self.try_swapping_bucket_path(temp_second_loc,
+                                                                    temp_first_loc)
+        if optimized_temp_second_loc is not None:
+            # TODO: Delete after
+            yield (temp_first_loc, optimized_temp_second_loc)
+            return
+        logger.warning("Could not match paths between source and destination buckets--"
+                       "will not be able to use Google Cloud Platform Data Transfer Service for "
+                       "cloud-based copy.")
+        yield (temp_first_loc, temp_second_loc)
+
+    @contextmanager
+    def _optimize_temp_second_location_for_gcp_data_transfer(self,
+                                                             permanent_first_loc: 'S3DirectoryUrl',
+                                                             temp_second_loc: 'GCSDirectoryUrl') ->\
+            Iterator[BaseDirectoryUrl]:
+        #
+        # GCP data transfer is great, but has the limitations:
+        #
+        # 1) It only works from GCS -> GCS or S3 -> GCS - like the
+        # roach motel, you can check in but you can't check out.
+        #
+        # 2) You can't specify a different destination location
+        # directory in the GCS bucket than in your source bucket
+        #
+        # So let's make sure that if at all possible, we use the same
+        # directory.
+        optimized_temp_second_loc = self.try_swapping_bucket_path(temp_second_loc,
+                                                                  permanent_first_loc)
+        if optimized_temp_second_loc is not None:
+            # TODO: Delete after
```

Comment:
Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker